dotnet/samples/Concepts/Memory/VectorStore_DataIngestion_CustomMapper.cs

```diff
@@ -0,0 +1,204 @@
+
+            var vector = (ReadOnlyMemory<float>)dataModel.Vectors["DefinitionEmbedding"];
+            var jsonArray = new JsonArray(vector.ToArray().Select(x => JsonValue.Create(x)).ToArray());
+            jsonObject.Add("DefinitionEmbedding", jsonArray);
+
+            return (dataModel.Key, jsonObject);
+        }
+
+        public GenericDataModel MapFromStorageToDataModel((string Key, JsonNode Node) storageModel, StorageToDataModelMapperOptions options)
+        {
+            var dataModel = new GenericDataModel
+            {
+                Key = storageModel.Key,
+                Data = new Dictionary<string, object>
+                {
+                    { "Term", (string)storageModel.Node["Term"]! },
+                    { "Definition", (string)storageModel.Node["Definition"]! }
+                },
+                Vectors = new Dictionary<string, object>
+                {
+                    { "DefinitionEmbedding", new ReadOnlyMemory<float>(storageModel.Node["DefinitionEmbedding"]!.AsArray().Select(x => (float)x!).ToArray()) }
+                }
+            };
+
+            return dataModel;
+        }
+    }
+
+    /// <summary>
+    /// A factory for creating collections in the vector store
+    /// </summary>
+    private sealed class Factory : IRedisVectorStoreRecordCollectionFactory
+    {
+        public IVectorStoreRecordCollection<TKey, TRecord> CreateVectorStoreRecordCollection<TKey, TRecord>(IDatabase database, string name, VectorStoreRecordDefinition? vectorStoreRecordDefinition)
+            where TKey : notnull
+            where TRecord : class
+        {
+            // If the record definition is the glossary definition and the record type is the generic data model, inject the custom mapper into the collection options.
+            if (vectorStoreRecordDefinition == s_glossaryDefinition && typeof(TRecord) == typeof(GenericDataModel))
+            {
+                var customCollection = new RedisJsonVectorStoreRecordCollection<GenericDataModel>(database, name, new() { VectorStoreRecordDefinition = vectorStoreRecordDefinition, JsonNodeCustomMapper = new Mapper() }) as IVectorStoreRecordCollection<TKey, TRecord>;
+                return customCollection!;
+            }
+
+            // Otherwise, just create a standard collection with the default mapper.
+            var collection = new RedisJsonVectorStoreRecordCollection<TRecord>(database, name, new() { VectorStoreRecordDefinition = vectorStoreRecordDefinition }) as IVectorStoreRecordCollection<TKey, TRecord>;
+            return collection!;
+        }
+    }
+
+    /// <summary>
+    /// Sample generic data model class that can store any data.
+    /// </summary>
+    private sealed class GenericDataModel
+    {
+        public string Key { get; set; }
+
+        public Dictionary<string, object> Data { get; set; }
+
+        public Dictionary<string, object> Vectors { get; set; }
+    }
```

Comment:
This is very generic model example for `Concepts` folder, should we use something closer to real life use-cases? I think `Hotel` model we are using in integration tests will work better. It will clearly show that as a developer, you can use your own model/schema as per your business case.

Reply:
In this case I'm actually trying to show that even though in storage you have a flat model with named properties, in the data model you can have this generic model.  I cannot really do it the other way around, since AzureAISearch and Redis doesn't support anything other than a flat structure.  And Qdrant has the same storage model as this data model, and our built in mapper for Qdrant already does mapping to a flat structure with named properties.

Ultimately, I think we need to turn this specific sample into built in functionality, where we add this data model to abstractions and where we have built in mappers for each vector store that can map to this model.  As a start though, I wanted to get a sample out, so we can demonstrate what's possible without any code changes.

Reply:
PS, supporting a model like this has come up quite a lot in conversation with various people, so I'm keen to demonstrate a sample with it for that reason too.

Reply:
Makes sense, thank you! My main concern was that such model is not strongly typed, but it looks like it should work for this scenario.